packages/nextcloud/lib/src/helpers/common.dart

```diff
@@ -1,2 +1,51 @@
-/// The result of a version check.
-typedef VersionSupported<T> = ({bool isSupported, T minimumVersion});
+import 'package:meta/meta.dart';
+import 'package:version/version.dart';
+
+/// Holds the [versions], [minimumVersion] and [maximumExclusiveVersion] of an app.
+@immutable
+class VersionCheck {
+  /// Creates a new [VersionCheck].
+  ///
+  /// If the [maximumExclusiveVersion] is `null` the compatibility of the major of the [minimumVersion] is checked.
+  const VersionCheck({
+    required this.versions,
+    required this.minimumVersion,
+    required this.maximumExclusiveVersion,
+  });
+
+  /// Current version of the app.
+  final List<Version>? versions;
+
+  /// Minimum version of the app.
+  final Version minimumVersion;
+
+  /// Maximum exclusive version of the app.
+  ///
+  /// This version is exclusive in the [isSupported] check.
+  final Version? maximumExclusiveVersion;
+
+  /// Whether the [versions] is allowed by the [minimumVersion] and [maximumExclusiveVersion].
+  ///
+  /// If the [maximumExclusiveVersion] is `null` the compatibility of the major of the [minimumVersion] is checked.
+  bool get isSupported {
+    if (versions == null || versions!.isEmpty) {
+      return true;
+    }
+
+    for (final version in versions!) {
+      if (version >= minimumVersion) {
+        if (maximumExclusiveVersion != null) {
+          if (version < maximumExclusiveVersion) {
+            return true;
+          }
+        } else {
+          if (version.major == minimumVersion.major) {
+            return true;
+          }
+        }
+      }
+    }
+
+    return false;
```

Comment:
This is still fairly complex. Why not supply a default max version in the constructor. Depending on my other question of how a list of versions should behave this could look:

```dart
  VersionCheck({
    required this.versions,
    required this.minimumVersion,
    required final Version? maximumExclusiveVersion,
  }) : maximumExclusiveVersion = maximumExclusiveVersion ?? minimumVersion.incrementMajor();

/* ... */

    for (final version in versions!) {
      if (version < minimumVersion || version > maximumExclusiveVersion) {
        return false;
      }
    }

    return true;
    
```